Draft - 8088998: LineChart: duplicate child added exception when remove & add a series

karthikpandelu opened the pull request:
While checking for duplicate series addition to the line chart, `setToRemove` value was not considered before throwing exception. Hence code to handling the case of adding the removed series was never run.

Added condition to check `setToRemove` boolean value before throwing exception. Made changes to call `setChart` method after calling `seriesBeingRemovedIsAdded`. Otherwise chart will not be drawn for the series, only points will be plotted.

This issue is reproducible only when animation is enabled because timeline will be still running when removed series is added back to the same chart. Since timeline does not run in unit tests, added system test to validate the fix.

kevinrushforth commented:
Reviewers: @aghaisas @andy-goryachev-oracle 

/reviewers 2

andy-goryachev-oracle commented:
@karthikpandelu are you going to fix the remaining charts as a part of https://bugs.openjdk.org/browse/JDK-8301445 or continue with this PR?

karthikpandelu commented:
I'll continue with this PR and fix the same issue seen in other charts.

karthikpandelu commented:
Fixed duplicate child added exception in all XYCharts.
In `StackedBarChart` Y axis value was being manipulated with relative zero position. Due to this the actual Y value was getting changed and chart was getting modified as we remove and add the series back to the chart. Fixed this issue as well.

Added system test to validate the fix for all the XYCharts.

@aghaisas and @andy-goryachev-oracle please review these changes.

andy-goryachev-oracle reviewed (changes requested):
looks like the main issue has been fixed, tested with MonkeyTester
https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/monkey/MonkeyTesterApp.java

A few minor things are noted in the comments, most important is whether it's possible to leave the Timeline running forever.

karthikpandelu commented on modules/javafx.controls/src/main/java/javafx/scene/chart/AreaChart.java:
```diff
@@ -71,6 +71,7 @@
 
     /** A multiplier for teh Y values that we store for each series, it is used to animate in a new series */
```
Fixed this typo

karthikpandelu commented on tests/system/src/test/java/test/javafx/scene/control/XYChartExceptionOnAddingRemovedSeriesTest.java:
```diff
@@ -0,0 +1,320 @@
+
+    @Test
+    public void testLineChartExceptionOnAddingRemovedSeries() throws Throwable {
+        Thread.sleep(1000); // Wait for stage to layout
```
Updated code to wait on `startupLatch` instead of sleep.

karthikpandelu commented on modules/javafx.controls/src/main/java/javafx/scene/chart/AreaChart.java:
```diff
@@ -554,6 +555,17 @@ LegendItem createLegendItemForSeries(Series<X, Y> series, int seriesIndex) {
+    @Override void seriesBeingRemovedIsAdded(Series<X,Y> series) {
+        if (timeline != null) {
+            timeline.setOnFinished(null);
+            timeline.stop();
```
> two questions:
> 
>     1. should the timeline object be set to null here?  the reference will get overwritten in seriesRemoved():414, there is probably no need to keep this object in memory
> 
Yes timeline can be made null here. Updated the code.

>     2. is it possible that `seriesBeingRemovedIsAdded()` will be invoked twice with a different `series` argument?  In other words, are we going to face a situation when a `timeline` from unrelated series will not get `stop`'ped?
> 
> 
> (these question apply to other charts as well)

Since `seriesBeingRemovedIsAdded()` is called from the event handler which gets invoked when the series changes and further there is a check on `setToRemove` boolean, I believe the method will not be called with different `series` and cause unwanted `series` timeline to be stopped.

andy-goryachev-oracle replied:
Thank you for clarification, @karthikpandelu

karthikpandelu commented on tests/system/src/test/java/test/javafx/scene/control/XYChartExceptionOnAddingRemovedSeriesTest.java:
```diff
@@ -0,0 +1,320 @@
+            stage.setOnShown(event -> Platform.runLater(startupLatch::countDown));
+            stage.show();
+
+            Thread.currentThread().setUncaughtExceptionHandler((t2, e) -> {
```
Updated code to set the handler in function annotated with @BeforeClass and cleaned up in  @AfterClass.

andy-goryachev-oracle replied:
I am not sure setting uncaught exception handler in @BeforeClass and clearing it in @AfterClass is correct.
The junit framework may execute these tests in parallel or sequentially (I think these are actually executed sequentially), but I would not rely on the fact that the individual tests will be executed in the same thread.

So I'd suggest to move this code to @After and @Before, just like BehaviorCleanupTest and other tests.

karthikpandelu replied:
Moved the code to @Before and @After. Please check

andy-goryachev-oracle reviewed (changes requested):
asking for one change.

andy-goryachev-oracle reviewed (approved):
looks good, thank you.

karthikpandelu commented:
/integrate

aghaisas commented:
/sponsor
